src/compiler/scala/tools/nsc/PhaseAssembly.scala

```diff
@@ -112,16 +112,19 @@ trait PhaseAssembly {
 
       if (node.level < lvl) node.level = lvl
 
-      var hls = Nil ++ node.before.filter(_.hard)
-      while (hls.size > 0) {
-        for (hl <- hls) {
-          node.phaseobj = Some(node.phaseobj.get ++ hl.frm.phaseobj.get)
-          node.before = hl.frm.before
-          nodes -= hl.frm.phasename
-          edges -= hl
-          for (edge <- node.before) edge.to = node
+      var befores = node.before
+      def hasHardLinks() = befores.exists(_.hard)
```

Comment:
This could be a flag `var sawHard` that is set on traversal. However, IIUC (which I probably don't), it has already verified that you can only have one hard dependency on a phase. So just `node.before.find(_.hard)`. You loop if the dependencies you're adding include a hard dependency (at most one).